### PR TITLE
ENH: Base implementation of B-Spline transforms

### PR DESCRIPTION
**Pull request by oesteban**
Addressing #106.

### CODE REVIEW

File: nitransforms/interp/bspline.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    shape = target_nii.shape[:3]
+    ctrl_sp = nb.affines.voxel_sizes(ctrl_nii.affine)[:3]
+    ras2ijk = np.linalg.inv(ctrl_nii.affine)
+    origin = nb.affines.apply_affine(ras2ijk, [tuple(target_nii.affine[:3, 3])])[0]
+
+    wd = []
+    for i, (o, n, sp) in enumerate(
+        zip(origin, shape, nb.affines.voxel_sizes(target_nii.affine)[:3])
+    ):
+        locations = np.arange(0, n, dtype="float16") * sp / ctrl_sp[i] + o
```
Comment (effigies) [inline nitransforms/interp/bspline.py:84]:
```suggestion
        # Locations of voxels in target image in control point image
        locations = np.arange(0, n, dtype="float16") * sp / ctrl_sp[i] + o
```

File: nitransforms/interp/bspline.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        A sparse matrix of interpolating weights :math:`\Psi^3(\mathbf{k}, \mathbf{s})`
+        for the *N* voxels of the target EPI, for each of the total *K* knots.
+        This sparse matrix can be directly used as design matrix for the fitting
+        step of approximation/extrapolation.
+
+    """
+    shape = target_nii.shape[:3]
+    ctrl_sp = nb.affines.voxel_sizes(ctrl_nii.affine)[:3]
+    ras2ijk = np.linalg.inv(ctrl_nii.affine)
+    origin = nb.affines.apply_affine(ras2ijk, [tuple(target_nii.affine[:3, 3])])[0]
```
Comment (effigies) [inline nitransforms/interp/bspline.py:78]: Mostly my own notes, but might help future readers.

```suggestion
    # IJK index in the control point image of the first index in the target image
    origin = nb.affines.apply_affine(ras2ijk, [tuple(target_nii.affine[:3, 3])])[0]
```

File: nitransforms/nonlinear.py [hunk 3]
```diff
@@ -33,13 +57,20 @@ def __init__(self, field, reference=None):
                 "the number of dimensions (%d)" % (self._field.shape[-1], ndim)
             )
 
-        self.reference = field.__class__(
-            np.zeros(self._field.shape[:-1]), field.affine, field.header
-        )
+    def __repr__(self):
+        """Beautify the python representation."""
+        return f"<{self._field.shape[:3]} field of {self._field.shape[-1]}D displacements>"
```
Comment (effigies) [inline nitransforms/nonlinear.py:62]:
```suggestion
        return f"<DisplacementFieldTransform[{self._field.shape[-1]}D] {self._field.shape[:3]}>"
```

File: nitransforms/nonlinear.py [hunk 4]
```diff
@@ -22,9 +31,24 @@ class DisplacementsFieldTransform(TransformBase):
...
 
     def __init__(self, field, reference=None):
-        """Create a dense deformation field transform."""
+        """
+        Create a dense deformation field transform.
+
+        Example
+        -------
+        >>> DisplacementsFieldTransform(test_dir / "someones_displacement_field.nii.gz")
+        <(57, 67, 56) field of 3D displacements>
```
Comment (effigies) [inline nitransforms/nonlinear.py:40]: Generally expect class name to come first.

```suggestion
        <DisplacementFieldTransform[3D] (57, 67, 56)>
```

File: nitransforms/nonlinear.py [hunk 5]
```diff
@@ -22,9 +31,24 @@ class DisplacementsFieldTransform(TransformBase):
...
+
+        """
         super().__init__()
-        self._field = np.asanyarray(field.dataobj)
+
+        field = _ensure_image(field)
+        self._field = np.squeeze(
+            np.asanyarray(field.dataobj) if hasattr(field, "dataobj") else field
+        )
```
Comment (effigies) [inline nitransforms/nonlinear.py:48]: No need to change anything, but I think `_ensure_image` doesn't quite capture the operation of something that will pass ndarrays through untouched.

File: nitransforms/nonlinear.py [hunk 6]
```diff
@@ -22,9 +31,24 @@ class DisplacementsFieldTransform(TransformBase):
...
+        self.reference = reference or field.__class__(
+            np.zeros(self._field.shape[:-1]), field.affine, field.header
```
Comment (effigies) [inline nitransforms/nonlinear.py:50]: This will fail if `field` is an `ndarray`.

File: nitransforms/nonlinear.py [hunk 7]
```diff
@@ -90,70 +118,162 @@ def from_filename(cls, filename, fmt="X5"):
...
+
+    def __init__(self, coefficients, reference=None, order=3):
+        """Create a smooth deformation field using B-Spline basis."""
+        super(BSplineFieldTransform, self).__init__()
+        self._order = order
+
+        coefficients = _ensure_image(coefficients)
+
+        self._coeffs = np.asanyarray(coefficients.dataobj)
+        self._knots = ImageGrid(four_to_three(coefficients)[0])
```
Comment (effigies) [inline nitransforms/nonlinear.py:135]: Seems wasteful to create a bunch of new images to take the first and throw it away almost immediately. ImageGrid already resets to `shape = shape[:3]` if `ndim == 4`.

```suggestion
        self._knots = ImageGrid(coefficients)
```

**Review by effigies — COMMENTED** [review]
Note to self: pick back up at `to_field()`.

File: nitransforms/nonlinear.py [hunk 8]
```diff
@@ -90,70 +118,162 @@ def from_filename(cls, filename, fmt="X5"):
...
+        reference = _ensure_image(reference)
+        _ref = self.reference if reference is None else SpatialReference.factory(reference)
+        if _ref is None:
+            raise ValueError("A reference must be defined")
+
+        ndim = self._coeffs.shape[-1]
+
+        # If locations to be interpolated are on a grid, use faster tensor-bspline calculation
+        if self._weights is None:
+            self._weights = grid_bspline_weights(_ref, self._knots)
```
Comment (effigies) [inline nitransforms/nonlinear.py:156]: Docstring says that `ctrl_nii` (`self._knots`) should be a SpatialImage, not an `ImageGrid`. It just needs `affine` and `shape`, so this is fine. Should just update the docstring.

File: nitransforms/interp/bspline.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+        ],
+    )
+
+
+def grid_bspline_weights(target_nii, ctrl_nii):
+    r"""
+    Evaluate tensor-product B-Spline weights on a grid.
+
+    For each of the *N* input samples :math:`(s_1, s_2, s_3)` and *K* control
+    points or *knots* :math:`\mathbf{k} =(k_1, k_2, k_3)`, the tensor-product
```
Comment (effigies) [inline nitransforms/interp/bspline.py:35]: This math seems off. In the equation, you have `s_i - k_i`, but nothing here constrains `N = K`.

File: nitransforms/nonlinear.py [hunk 10]
```diff
@@ -90,70 +118,162 @@ def from_filename(cls, filename, fmt="X5"):
...
+
+        field = np.zeros((_ref.npoints, ndim))
+
+        for d in range(ndim):
+            field[:, d] = self._coeffs[..., d].reshape(-1) @ self._weights
```
Comment (effigies) [inline nitransforms/nonlinear.py:161]: Mostly a note to myself... it's hard to keep track of these

```suggestion
            #  1 x Nvox :                          (1 x K) @ (K x Nvox)
            field[:, d] = self._coeffs[..., d].reshape(-1) @ self._weights
```

File: nitransforms/nonlinear.py [hunk 11]
```diff
@@ -90,70 +118,162 @@ def from_filename(cls, filename, fmt="X5"):
...
+            output_dtype=output_dtype,
+        )
+
+    def map(self, x, inverse=False):
+        r"""
+        Apply the transformation to a list of physical coordinate points.
+
+        .. math::
+            \mathbf{y} = \mathbf{x} + \Psi^3(\mathbf{k}, \mathbf{x}),
+            \label{eq:2}\tag{2}
```
Comment (effigies) [inline nitransforms/nonlinear.py:219]:
```suggestion
            \label{eq:1}\tag{1}
```

File: nitransforms/nonlinear.py [hunk 12]
```diff
@@ -90,70 +118,162 @@ def from_filename(cls, filename, fmt="X5"):
...
+        # If locations to be interpolated are on a grid, use faster tensor-bspline calculation
+        if self._weights is None:
+            self._weights = grid_bspline_weights(_ref, self._knots)
+
+        field = np.zeros((_ref.npoints, ndim))
+
+        for d in range(ndim):
+            field[:, d] = self._coeffs[..., d].reshape(-1) @ self._weights
+
+        return field.astype(dtype)
```
Comment (effigies) [inline nitransforms/nonlinear.py:163]: Seems more sensible to wrap this field with metadata for interpretation now, than to make the caller do it.

```suggestion
        return DisplacementsFieldTransform(
            field.astype(dtype).reshape(*_ref.shape, -1), reference=_ref)
```

File: nitransforms/nonlinear.py [hunk 13]
```diff
@@ -90,70 +118,162 @@ def from_filename(cls, filename, fmt="X5"):
...
+                cval=cval,
+                prefilter=prefilter,
+                output_dtype=output_dtype,
+            )
+
+        # If locations to be interpolated are on a grid, generate a displacements field
+        return DisplacementsFieldTransform(
+            self.to_field().reshape((*(_ref.shape), -1)),
+            reference=_ref,
+        ).apply(
```
Comment (effigies) [inline nitransforms/nonlinear.py:203]:
```suggestion
        return self.to_field().apply(
```

**Review by effigies — APPROVED** [review]
Based on your comment, I didn't really try to understand `_map_xyz`. The remaining parts seem sensible. Some final comments.

File: nitransforms/nonlinear.py (hunk 6 above)
Comment (oesteban) [reply nitransforms/nonlinear.py:50]: Yes, perhaps this should be more explicit that field must be an image if reference is `None`. Will add an error

File: nitransforms/nonlinear.py (hunk 12 above)
Comment (oesteban) [reply nitransforms/nonlinear.py:163]: The problem is that `_ref` might be `None` and it should be possible to still get the displacements.

File: nitransforms/nonlinear.py (hunk 5 above)
Comment (oesteban) [reply nitransforms/nonlinear.py:48]: Is this a comment on the function's name, the function's behavior, or both? 
Happy to move towards something more appropriate.

File: nitransforms/interp/bspline.py (hunk 9 above)
Comment (oesteban) [reply nitransforms/interp/bspline.py:35]: *i* indexes the dimension (therefore 1, 2, 3 subscripts). *N* is the number of voxels (reference.dataobj.size) and *K* the number of control points (knots.dataobj.size). I have deeply re-written the equation to make it clearer.

**Pull request merged**